homeassistant/components/proximity/sensor.py

```diff
@@ -0,0 +1,120 @@
+
+from .const import ATTR_DIR_OF_TRAVEL, ATTR_DIST_TO, ATTR_NEAREST, DOMAIN
+from .coordinator import ProximityDataUpdateCoordinator
+
+SENSORS_PER_ENTITY: list[SensorEntityDescription] = [
+    SensorEntityDescription(
+        key=ATTR_DIST_TO,
+        device_class=SensorDeviceClass.DISTANCE,
+        native_unit_of_measurement=UnitOfLength.METERS,
+    ),
+    SensorEntityDescription(
+        key=ATTR_DIR_OF_TRAVEL,
+        translation_key=ATTR_DIR_OF_TRAVEL,
+        icon="mdi:compass-outline",
+    ),
+]
+
+SENSORS_PER_PROXIMITY: list[SensorEntityDescription] = [
+    SensorEntityDescription(
+        key=ATTR_NEAREST, translation_key=ATTR_NEAREST, icon="mdi:near-me"
+    ),
+]
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up the Proximity sensor platform."""
+    if discovery_info is None:
+        return
+
+    coordinator: ProximityDataUpdateCoordinator = hass.data[DOMAIN][
+        discovery_info[CONF_NAME]
+    ]
+
+    entities: list[ProximitySensor | ProximityTrackedEntitySensor] = [
+        ProximitySensor(sensor, coordinator) for sensor in SENSORS_PER_PROXIMITY
+    ]
+
+    entities += [
+        ProximityTrackedEntitySensor(sensor, coordinator, tracked_entity)
+        for sensor in SENSORS_PER_ENTITY
+        for tracked_entity in coordinator.proximity_devices
+    ]
+
+    async_add_entities(entities)
+
+
+class ProximitySensor(SensorEntity, CoordinatorEntity[ProximityDataUpdateCoordinator]):
+    """Represents a Proximity sensor."""
+
+    def __init__(
+        self,
+        description: SensorEntityDescription,
+        coordinator: ProximityDataUpdateCoordinator,
+    ) -> None:
+        """Initialize the proximity."""
+        super().__init__(coordinator)
+
+        self.entity_description = description
+
+        self._attr_unique_id = slugify(
+            f"{slugify(coordinator.friendly_name)}_{description.key}"
+        )
+
+    @property
+    def native_value(self) -> str | float | None:
+        """Return native sensor value."""
+        if (
+            value := self.coordinator.data.proximity[self.entity_description.key]
+        ) == "not set":
+            return None
+        return value
+
+
+class ProximityTrackedEntitySensor(
+    SensorEntity, CoordinatorEntity[ProximityDataUpdateCoordinator]
+):
+    """Represents a Proximity tracked entity sensor."""
+
+    def __init__(
+        self,
+        description: SensorEntityDescription,
+        coordinator: ProximityDataUpdateCoordinator,
+        tracked_entity: str,
+    ) -> None:
+        """Initialize the proximity."""
+        super().__init__(coordinator)
+
+        self.entity_description = description
+        self.tracked_entity = tracked_entity
+
+        self._attr_unique_id = slugify(
+            f"{slugify(coordinator.friendly_name)}_{tracked_entity}_{description.key}"
```

Comment:
> There's two slugify calls?

fixed 👍 

> What happens if the tracked entity_id changes?
>
> Including a name in the unique_id isn't good. When we allow the user to set the unique_id it is an explicit setting for unique_id. If we take the name the unique_id setting isn't clear anymore to the user.

the user sets explicit the tracked entity ids in the config (_later via config flow_) so they cannot change by accident (_eq. user renames the entity_), but explizit by the user when changing the proximity configuration.

> If we're going to add a config entry to this integration I suggest we wait with adding unique_id until we have a config flow. When we have that we can use the config entry id as unique id.

i don't think this will work, because we can have multiple tracked entity ids per config entry, so config entry id would still not be unique. SO i'm not sure what unique source for a tracked entity we could use (_maybe the unique_id of the tracked entity - in case there is one_) 🤔

Reply:
My point is that it's not clear that the user is setting a unique_id when they set the zone to be monitored.

I meant that we can use the config entry id instead of the name of the zone. We'd still need to append the entity_id of the tracked entity, or some other id for each tracked entity, yes.

Yes, the tracked entity cannot be changed implicitly, but what should happen when the user changes the tracked entities? Should the stale sensor entities be cleaned up? Should sensor customizations be migrated to the new sensor?

Reply:
> My point is that it's not clear that the user is setting a unique_id when they set the zone to be monitored.

ah ... i misunderstood it (_I thought you meant the id of the tracked entities_).
The `coordinator.friendly_name` is not the monitored zone, but just a name the user gives this particular configuration (_see [`friendly_name`](https://www.home-assistant.io/integrations/proximity/#friendly_name) parameter in the docs_)
But sure, we can wait for the configflow PR and add unique_ids there 👍 
---

> Yes, the tracked entity cannot be changed implicitly, but what should happen when the user changes the tracked entities? Should the stale sensor entities be cleaned up? Should sensor customizations be migrated to the new sensor? Or should we not allow changing tracked entities for a config entry?

The idea is to make at least the list of tracked entities adjustable via an additional optionsflow.
When the user changes this list, new sensor entities will be created, but yes we should think about an automatic cleanup of stale sensors in this case - maybe as an improvement in an upcoming PR 🤔 
Further when the user renames an entity_id via the UI the user is warned like:

> Do you also want to rename the entity IDs of your entities?
> This will not change any configuration (like automations, scripts, scenes, dashboards) that is currently using these entities! You will have to update them yourself to use the new entity IDs!

Reply:
unique_id remove --> 3647395